internal/pkg/gateway/api.go

```diff
@@ -79,32 +95,47 @@ func (gs *Server) Endorse(ctx context.Context, proposedTransaction *gp.ProposedT
 	// send to all the endorsers
 	for _, e := range endorsers {
 		wg.Add(1)
-		go func(endorser *endorser) {
+		go func(e *endorser) {
 			defer wg.Done()
-			response, err := endorser.client.ProcessProposal(ctx, signedProposal)
-			responseCh <- &endorserResponse{pr: response, err: err}
+			response, err := e.client.ProcessProposal(ctx, signedProposal)
+			var detail *gp.EndpointError
+			if err != nil {
+				detail = &gp.EndpointError{Address: e.address, MspId: e.mspid, Message: err.Error()}
+				responseCh <- &endorserResponse{err: detail}
+				return
+			}
+			if response.Response.Status < 200 || response.Response.Status >= 400 {
+				// this is an error case and will be returned in the error details to the client
+				detail = &gp.EndpointError{Address: e.address, MspId: e.mspid, Message: fmt.Sprintf("error %d, %s", response.Response.Status, response.Response.Message)}
+			}
+			responseCh <- &endorserResponse{pr: response, err: detail}
```

Comment:
This could be easer to read with a helper and a switch. Can you please address this in a follow-up?

```go
response, err := e.client.ProcessProposal(ctx, signedProposal)
switch {
    case err != nil:
        responseCh <- &endorserResponse{err: endpointError(e, err)}
    case response.Response.Status < 200 || response.Response.Status >= 400:
        responseCh <- &endorserResponse{err: endpointError(e, fmt.Errorf("...", ...))
    default:
        responseCh <- &endorserResponse{pr: response}
}
```